barretenberg/cpp/src/barretenberg/client_ivc/client_ivc.cpp

```diff
@@ -213,40 +213,103 @@ void ClientIVC::accumulate(ClientCircuit& circuit, const std::shared_ptr<Verific
+ * @return HonkProof - a Mega proof
  */
-ClientIVC::Proof ClientIVC::prove()
+HonkProof ClientIVC::construct_and_prove_hiding_circuit()
 {
     max_block_size_tracker.print();               // print minimum structured sizes for each block
     ASSERT(verification_queue.size() == 1);       // ensure only a single fold proof remains in the queue
     ASSERT(merge_verification_queue.size() == 1); // ensure only a single merge proof remains in the queue
+
     FoldProof& fold_proof = verification_queue[0].proof;
-    MergeProof& merge_proof = merge_verification_queue[0];
     HonkProof decider_proof = decider_prove();
-    // Free the accumulator to save memory
+
     fold_output.accumulator = nullptr;
-    return { fold_proof, std::move(decider_proof), goblin.prove(merge_proof) };
+
+    ClientCircuit builder{ goblin.op_queue };
+    // The last circuit being folded is a kernel circuit whose public inputs need to be passed to the base rollup
+    // circuit. So, these have to be preserved as public inputs to the hiding circuit (and, subsequently, as public
+    // inputs to the tube circuit) which are intermediate stages.
+    // TODO(https://github.com/AztecProtocol/barretenberg/issues/1048): link these properly, likely insecure
+    auto num_public_inputs = static_cast<uint32_t>(static_cast<uint256_t>(fold_proof[1]));
+    vinfo("num_public_inputs of the last folding proof BEFORE SUBTRACTION", num_public_inputs);
+    num_public_inputs -= bb::AGGREGATION_OBJECT_SIZE;             // exclude aggregation object
+    num_public_inputs -= bb::PROPAGATED_DATABUS_COMMITMENTS_SIZE; // exclude propagated databus commitments
+    vinfo("num_public_inputs of the last folding proof ", num_public_inputs);
+    for (size_t i = 0; i < num_public_inputs; i++) {
+        size_t offset = 3;
+        builder.add_public_variable(fold_proof[i + offset]);
+    }
+
+    goblin.verify_merge(builder, merge_verification_queue[0]);
+    merge_verification_queue.clear();
+
+    // Construct stdlib accumulator, decider vkey and folding proof
+    auto stdlib_verifier_accumulator =
+        std::make_shared<RecursiveDeciderVerificationKey>(&builder, verifier_accumulator);
+
+    auto stdlib_decider_vk =
+        std::make_shared<RecursiveVerificationKey>(&builder, verification_queue[0].honk_verification_key);
+
+    auto stdlib_proof = bb::convert_proof_to_witness(&builder, fold_proof);
+
+    // Perform recursive folding verification oft he la
+    FoldingRecursiveVerifier folding_verifier{ &builder, stdlib_verifier_accumulator, { stdlib_decider_vk } };
+    auto recursive_verifier_accumulator = folding_verifier.verify_folding_proof(stdlib_proof);
+    auto native_verifier_acc = std::make_shared<DeciderVerificationKey>(recursive_verifier_accumulator->get_value());
+    verification_queue.clear();
+
+    // Perform recursive decider verification
+    DeciderRecursiveVerifier decider{ &builder, native_verifier_acc };
```

Comment:
Does this not have a constructor from a stdlib accumulator? Seems crazy to convert the accum to native then back to stdlib internally here. Should be easy enough to add such a constructor if it doesnt exist

Reply:
no, it doesn't, but it's a good point that one should exist. I think when I implemented these back in February we didn't consider the recursive folding verifier and decider verifier sharing the same builder.